### DISCUSSION TITLE
Accessing Response Body inside of middleware throwing error `Too little data for declared Content-Length`

### DISCUSSION THREAD

**Discussion by McHulotte** [fastapi/fastapi#11064] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# I used app.add_middleware() instead of annotation
async def log_requests(request: Request, call_next):
    # Some code before receiving response
    response = await call_next(request)
    if response.status_code == 200:
        if request.url.path == '/token':
            response_body = b""
            async for chunk in response.body_iterator:
                print(chunk)
    return response
```

### Description

Hi! Here is my problem,

When I try to read the response body in my `log_requests` function which is the `Middleware` I get this error:
`h11._util.LocalProtocolError: Too little data for declared Content-Length`
I understand that this may have to do with the response body that isn't the same length as what is stored in the Content-Length in headers.
However, my goal isn't to modify the body of the response, it's simply to access it and read it. So my code shouldn't be modifying the length of it.
My code works completely fine without this part : 
```python
async for chunk in response.body_iterator:
                print(chunk)
```

So can someone please tell me what I am doing wrong and why accessing the response body is throwing me this error.
(I already tried comparing the content length and the actual length of the body and it's the same with and without the `body_iterator used`)

Thanks in advance,

Regards

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

1.10.12

### Python Version

Python 3.10.12

**Comment by VatsalJagani** [comment]
@McHulotte - This is one of the major problems with Fastapi and Starlette and the request and response body. As this will clean up the body as soon as you read it. So once you read it there is no body anymore which does not match with size of the body in the response header.

Generalize problem here - https://github.com/tiangolo/fastapi/discussions/11330

**Comment by YuriiMotov** [comment, marked as answer]
As far as I know, conventional approach is to write [Pure ASGI Middleware](https://www.starlette.io/middleware/#pure-asgi-middleware) and wrap `send` generator to add your logging logic.

You can look it up here: https://github.com/Kludex/asgi-logger/blob/3247a1ac80658b3e42880f91eaeb02065ce8f1b1/asgi_logger/middleware.py#L22-L67
Not sure that code still works (last updated 4 years ago), but general approach should be the same
